Framework/HistogramData/src/Interpolate.cpp

```diff
@@ -94,103 +103,249 @@ void sanityCheck(const Histogram &input, const Histogram &output,
-                        YInterp &outY, const InterpolationType type) {
-  const gsl_interp_type *interpType = nullptr;
-  switch (type) {
-  case InterpolationType::LINEAR:
-    interpType = gsl_interp_linear;
-    break;
-  case InterpolationType::CSPLINE:
-    interpType = gsl_interp_cspline;
-    break;
+void interpolateYCSplineInplace(const Mantid::HistogramData::Histogram &input,
+                                const Mantid::HistogramData::Points &points,
+                                Mantid::HistogramData::Histogram &output,
+                                const bool calculateErrors = false,
+                                const bool independentErrors = true) {
+  auto xs = input.dataX();
+  // create tridiagonal "h" matrix
+  Mantid::Kernel::Matrix<double> h(xs.size() - 2, xs.size() - 2);
+  for (size_t i = 0; i < xs.size() - 2; i++) {
+    for (size_t j = 0; j < xs.size() - 2; j++) {
+      if (i == j) {
+        h[i][j] = (xs[i + 2] - xs[i]) / 3;
+      }
+      if (i == j + 1) {
+        h[i][j] = (xs[i] - xs[j]) / 6;
+      }
+      if (j == i + 1) {
+        h[i][j] = (xs[i + 2] - xs[i + 1]) / 6;
+      }
+    }
   }
-  using gsl_interp_uptr = std::unique_ptr<gsl_interp, void (*)(gsl_interp *)>;
-  auto interp =
-      gsl_interp_uptr(gsl_interp_alloc(interpType, xs.size()), gsl_interp_free);
-  gsl_interp_init(interp.get(), xs.data(), ys.data(), xs.size());
-  using gsl_interp_accel_uptr =
-      std::unique_ptr<gsl_interp_accel, void (*)(gsl_interp_accel *)>;
-  auto lookupTable =
-      gsl_interp_accel_uptr(gsl_interp_accel_alloc(), gsl_interp_accel_free);
-  // Evaluate each point for the full range
-  for (size_t i = 0; i < outY.size(); ++i) {
-    outY[i] = gsl_interp_eval(interp.get(), xs.data(), ys.data(), points[i],
-                              lookupTable.get());
+  std::vector<double> d(xs.size() - 2);
+  auto ys = input.dataY();
+  for (size_t i = 0; i < xs.size() - 2; i++) {
+    d[i] = (ys[i + 2] - ys[i + 1]) / (xs[i + 2] - xs[i + 1]) -
+           (ys[i + 1] - ys[i]) / (xs[i + 1] - xs[i]);
+  }
+
+  // ypp means y prime prime
+  std::vector<double> ypp(xs.size() - 2);
+  // would be quicker to solve linear equation rather than invert h but also
+  // need h-1 elements later on
+  h.Invert();
+  ypp = h * d;
+
+  // add in the zero second derivatives at extreme pts to give natural splines
+  std::vector<double> ypp_full(xs.size(), 0);
+  std::copy(ypp.begin(), ypp.end(), ypp_full.begin() + 1);
+
+  // calculate some covariances to support error propagation
+  auto &enew = output.mutableE();
+  auto &eold = input.dataE();
+  std::vector<double> u_ypp_ypp(xs.size());
+  std::vector<double> u_ypp_y(xs.size());
+
+  for (size_t i = 0; i < xs.size(); i++) {
+    for (size_t k = 0; k < xs.size(); k++) {
+      double dyppidyk = 0;
+      if ((i != 0) && (i != xs.size() - 1)) {
+        if (k > 1) {
+          dyppidyk += h[i - 1][k - 2] / (xs[k] - xs[k - 1]);
```

Comment:
These variable names could do with being more clear what they are.

Reply:
Yes I'll see what I can do. I'll probably add a few comments to explain my naming convention eg dyppidyk is meant to be dy''_i/dy_k but was bit hindered by not being able to include certain characters in the variable names. So have used p to mean "prime"